### PR TITLE
Symbol tutorial improvements

### PR DESCRIPTION
**Pull request by Roshrini**
@zackchase @madjam @nswamy Please take a look.

### CODE REVIEW

**Comment by piiswrong** [comment]
I'll merge this first but we shouldn't consider this tutorial "done". A lot of contents are outdated/inaccurate.

For example the variable sharing section is wrong. The right way to do it is to use a single variable: b = a + a*a.

Also calculating derivatives is not the advantage of symbolic api. Pytorch does it with imperative autograd. In fact we also have imperative autograd and the higher level interface will be there shortly.

The only benefit of symbolic API is runtime and memory optimizations. The explanation on why this is the case and how we are doing it is also inaccurate. The intro seems a little verbose for people who just want to know how to use it. Its probably better to briefly state what's the benefit and refer users to design notes for the why & hows.

File: docs/tutorials/basic/symbol.md [hunk 1]
```diff
@@ -1,23 +1,73 @@
...
+`NDArray` allows you to write programs for scientific computation
+in an imperative fashion using, making full use of the native control of any front end language.
+So you might wonder, why don't we just use `NDArray` for all computation?
+
+MXNet also provides the Symbol API, an interface for symbolic programming.
+With symbolic programing, rather than executing operations step by step,
+we first define a *computation graph*.
+This graph contains placeholders for inputs and designated outputs.
+We can then compile the graph, yielding a function
+that can be bound to real data and run.
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: real data -> 'NDArray' ?

File: docs/tutorials/basic/symbol.md [hunk 2]
```diff
@@ -1,23 +1,73 @@
...
+MXNet's Symbol API is similar to the network configurations
+used by [Caffe](http://caffe.berkeleyvision.org/)
+and [CXXNet](https://github.com/dmlc/cxxnet)
+and the symbolic programming in [Theano](http://deeplearning.net/software/theano/).
+
+There are a few reasons for running symbolic operations.
+First, building symbolic computation graphs
+makes it easy to automatically calculate derivatives.
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: Calucating derivative is not unique to symbolic graph. Pytorch does it with imperative autograd. We have it as an experimental feature too.

File: docs/tutorials/basic/symbol.md [hunk 3]
```diff
@@ -1,23 +1,73 @@
...
+to taking derivatives (the *gradient*) of the objective
+with respect to the parameters.
+By updating the parameters in the direction indicated by the gradient,
+we improve performance with respect to the objective.
+Because taking derivatives is so integral to training neural networks,
+the ability to compute them automatically and efficiently is a tremendous advantage.
+
+Another advantage conferred by symbolic aproach is that
+we can optimize our functions before using them.
+For example, we execute mathematical computations in imperative fashion,
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: *when* we?

File: docs/tutorials/basic/symbol.md [hunk 4]
```diff
@@ -1,23 +1,73 @@
...
+we don't know at the time that we run each operation,
+which values will be needed later on.
+But with symbolic programming, we declare the required outputs in advance.
+This means that we can recycle memory allocated in intermediate steps,
+as by performing operations in place.
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: How about a brief sentence saying symbolic API uses less memory for the same network and refer users to design notes for how and why.

File: docs/tutorials/basic/symbol.md [hunk 5]
```diff
@@ -1,23 +1,73 @@
...
+
+In our design notes, we present [a more thorough disussion on the comparative strengths
+of imperative and symbolic programing](http://mxnet.io/architecture/program_model.html).
+But in this document, we'll focus on teaching you how to use MXNet's Symbol API.
+In MXNet, we can compose Symbols from other Symbols, using operators,
+such as simple matrix operations (e.g. "+"),
+or whole neural network layers (e.g. convolution layer).
+Operator can take multiple input variables,
+can produce multiple output variables
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: Only inputs are called variables. internal and output nodes are symbols. variable is a special case of symbol

File: docs/tutorials/basic/symbol.md [hunk 6]
```diff
@@ -1,23 +1,73 @@
...
-symbol.
-
-## Symbol Composition
+In a [previous tutorial](ndarray.md), we introduced `NDArray`,
+the basic data structure for manipulating data in MXNet.
+And just using NDArray by itself, we can execute a wide range of mathematical operations.
+In fact, we could define and update a full neural network just by using `NDArray`.
+`NDArray` allows you to write programs for scientific computation
+in an imperative fashion using, making full use of the native control of any front end language.
+So you might wonder, why don't we just use `NDArray` for all computation?
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: Our imperative autograd interface is in progress and will be public soonish. At that time we should refactor the doc to start with a page that help users to decide whether they should use symbolic or imperative API. Then 2 page for symbolic and imperative.

File: docs/tutorials/basic/symbol.md [hunk 7]
```diff
@@ -1,23 +1,73 @@
...
+In our design notes, we present [a more thorough disussion on the comparative strengths
+of imperative and symbolic programing](http://mxnet.io/architecture/program_model.html).
+But in this document, we'll focus on teaching you how to use MXNet's Symbol API.
+In MXNet, we can compose Symbols from other Symbols, using operators,
+such as simple matrix operations (e.g. "+"),
+or whole neural network layers (e.g. convolution layer).
+Operator can take multiple input variables,
+can produce multiple output variables
+and can maintain internal state variables.
+Variables can be either free, bound to real data, or represent output of another symbol.
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: real data -> ndarray?

File: docs/tutorials/basic/symbol.md [hunk 8]
```diff
@@ -1,23 +1,73 @@
...
-The following example composites a simple expression `a+b`. We first create the
-placeholders `a` and `b` with names using `mx.sym.Variable`, and then construct
-the desired symbol by using the operator `+`. When the string name is not given
-during creating, MXNet will automatically generate a unique name for the symbol,
-which is the case for `c`.
+The following example builds a simple expression: `a+b`.
+First, we create two placeholders with  `mx.sym.Variable`,
+giving them the names `a` and `b`.
+We then construct the desired symbol by using the operator `+`.
+If we don't name our variables while creating them,
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: don't need to name our *operators*

File: docs/tutorials/basic/symbol.md [hunk 9]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
 mx.viz.plot_network(net, shape={'data':(100,200)})
 ```
 
-### Modulelized Construction for Deep Networks
+Each symbol takes a (unique) string name.
+*Variable* typically represents inputs to the computation graph.
+Symbols defined via operators typically take other symbols as their input data,
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: same as above. The concepts are confused here. Symbols corresponds to NDArray. They both represent a single tensor. Operators represent the computation between tensors. Operators take symbol (or NDArray) as inputs and produce it as output.

File: docs/tutorials/basic/symbol.md [hunk 10]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
...
+such as the number of hidden neurons (*num_hidden*) or the activation type (*act_type*).
+
+We can view a symbol simply as a function taking several arguments.
+And we can retrieve those arguments with the following method call:
+
+```python
+net.list_arguments()
+```
+
+These arguments are the parameters needed by each symbol:
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: parameters & inputs

File: docs/tutorials/basic/symbol.md [hunk 11]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
...
+
+- *data*: Input data needed by the variable *data*.
+- *fc1_weight* and *fc1_bias*: The weight and bias for the first fully connected layer *fc1*.
+- *fc2_weight* and *fc2_bias*: The weight and bias for the second fully connected layer *fc2*.
+- *out_label*: The label needed by the loss.
+
+We can also specify the automatically generated names explicitly:
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: remove " automatically generated" ?

File: docs/tutorials/basic/symbol.md [hunk 12]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
...
+MXNet provides well-optimized symbols for layers commonly used in deep learning
+(see [src/operator](https://github.com/dmlc/mxnet/tree/master/src/operator)).
+We can also define new operators in Python. The following example first
+performs an element-wise add between two symbols, then feeds them to the fully
+connected operator:
+
+```python
+lhs = mx.symbol.Variable('data1')
+rhs = mx.symbol.Variable('data2')
+net = mx.symbol.FullyConnected(data=lhs + rhs, name='fc1', num_hidden=128)
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: explain somewhere that FullyConnected, for example, has 3 inputs, data, weight, bias. When you don't specify an input, an variable will be automatically generated for it.

File: docs/tutorials/basic/symbol.md [hunk 13]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
...
+```
+
+We can also construct a symbol in a more flexible way than the single forward
+composition depicted in the preceding example:
+
+```python
+net = mx.symbol.Variable('data')
+net = mx.symbol.FullyConnected(data=net, name='fc1', num_hidden=128)
+net2 = mx.symbol.Variable('data2')
+net2 = mx.symbol.FullyConnected(data=net2, name='net2', num_hidden=128)
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: name='fc2'

File: docs/tutorials/basic/symbol.md [hunk 14]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
...
+composed_net = net(data=net2, name='compose')
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: composed_net = net2(data=net1, name='compose')

File: docs/tutorials/basic/symbol.md [hunk 15]
```diff
@@ -56,60 +111,150 @@ net = mx.sym.SoftmaxOutput(data=net, name='out')
...
+                      pad=(1, 1), name=('%s_double_3x3_0' % name))
+    cd3x3 = ConvFactory(data=cd3x3, num_filter=num_d3x3, kernel=(3, 3),
+                      pad=(1, 1), name=('%s_double_3x3_1' % name))
     # pool + proj
-    pooling = mx.sym.Pooling(data=data, kernel=(3, 3), stride=(1, 1), pad=(1, 1), pool_type=pool, name=('%s_pool_%s_pool' % (pool, name)))
-    cproj = ConvFactory(data=pooling, num_filter=proj, kernel=(1, 1), name=('%s_proj' %  name))
+    pooling = mx.sym.Pooling(data=data, kernel=(3, 3), stride=(1, 1), pad=(1, 1),
+                      pool_type=pool, name=('%s_pool_%s_pool' % (pool, name)))
+    cproj = ConvFactory(data=pooling, num_filter=proj, kernel=(1, 1),
+                      name=('%s_proj' %  name))
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: indentations are wrong for these lines.
should be
```
ConvFactory(...
                      name
```

File: docs/tutorials/basic/symbol.md [hunk 16]
```diff
@@ -123,66 +268,70 @@ group.list_outputs()
...
-- provides a large number of neural network related operators such as
-  Convolution, Activation, and BatchNorm
-- provides automatic differentiation
-- easy to construct and manipulate complex computations such as deep neural
-  networks
-- easy to save, load, and visualization
-- easy for the backend to optimize the computation and memory usage
+- Provides almost all functionalities of NDArray, such as +, \*, sin,
+  reshape etc.
+- Provides a large number of neural network related operators such as
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: not true. ndarray has these too.

File: docs/tutorials/basic/symbol.md [hunk 17]
```diff
@@ -123,66 +268,70 @@ group.list_outputs()
...
+  Convolution, Activation and BatchNorm.
+- Provides automatic differentiation.
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: ndarray has imperative autograd now too.

File: docs/tutorials/basic/symbol.md [hunk 18]
```diff
@@ -123,66 +268,70 @@ group.list_outputs()
...
+- Easy to construct and manipulate complex computations such as deep neural
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: not particularly easier than ndarray.

File: docs/tutorials/basic/symbol.md [hunk 19]
```diff
@@ -123,66 +268,70 @@ group.list_outputs()
...
 SQL.
 
 The pros for `NDArray`:
 
-- straightforward
-- easy to work with other language features (for loop, if-else condition, ..)
-  and libraries (numpy, ..)
-- easy to step-by-step debug
+- Straightforward.
+- Easy to work with other language features (for loop, if-else condition, ..)
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: native language features

File: docs/tutorials/basic/symbol.md [hunk 20]
```diff
@@ -201,12 +350,25 @@ ex_gpu.forward()
...
+creates all of the argument arrays for you. Then you can call ```forward```,
+and ```backward``` (if the gradient is needed) to get the gradient.
+
 ### Load and Save
 
-Similar to NDArray, we can either serialize a `Symbol` object by using `pickle`,
-or use `save` and `load` directly. Different to the binary format chosen by
-`NDArray`, `Symbol` uses the more readable json format for serialization. The
-`tojson` method returns the json string.
+Similar to `NDArray`, we can either serialize a `Symbol` object by using `pickle`,
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: We need to consider introducing the concept of *Graph*. 
Logically Symbols correspond to NDArray; they both represent a tensor. They are both outputs/inputs of operators. Graph is composed by chaining operators. We don't have a Graph object in frontend (but we do have graph in backend) Graphs are implicitly represented by output symbols.

File: docs/tutorials/basic/symbol.md [hunk 21]
```diff
@@ -217,24 +379,28 @@ c.tojson() == c2.tojson()
...
-`NDArray` is the default type of arguments in both `forward` and
-`backward`. Therefore we often also implement the computation with `NDArray`
-operations. To show the flexibility of MXNet, however, we will demonstrate an
-implementation of the `softmax` layer using NumPy. Though a NumPy based operator
-can be only run on CPU and also lose some optimizations which can be applied on
-NDArray, it enjoys the rich functionalities provided by NumPy.
-
-We first create a subclass of `mx.operator.CustomOp` and then define `forward`
-and `backward`.
+Most operators such as `mx.sym.Convolution` and `mx.sym.Reshape`
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: are we duplicating the new operator guide here? Consider putting a link instead?

File: docs/tutorials/basic/symbol.md [hunk 22]
```diff
@@ -293,12 +460,15 @@ net = mx.sym.Custom(data=prev_input, op_type='softmax')
...
+By default, MXNet uses 32-bit floats.
+But for better accuracy-performance,
+we can also use a lower precision data type.
+For example, The Nvidia Tesla Pascal GPUs
+(e.g. P100) have improved 16-bit float performance,
+while GTX Pascal GPUs (e.g. GTX 1080) are fast on 8-bit integers.
 
-We can use the `mx.sym.Cast` operator to convert the data type.
+To convert the data type as per the requirements,
+we can use `mx.sym.Cast` operator as follows:
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: we use cast now. lower case.

File: docs/tutorials/basic/symbol.md [hunk 23]
```diff
@@ -328,4 +498,20 @@ ex.forward()
...
+## How Efficient Is the Symbolic API?
+
+In short, it is designed to be very efficient in both memory and runtime.
+The major reason for introducing the Symbolic API is to bring the efficient C++
+operations in powerful toolkits, such as CXXNet and Caffe, together with the
+flexible dynamic NDArray operations. To maximize runtime performance and memory
+utilization, all of the memory and computation resources are allocated
+statically during the bind operation.
+
+The coarse-grained operators are equivalent to CXXNet layers,
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: people rarely know cxxnet. Consider removing mention of it and use caffe instead?

File: docs/tutorials/basic/symbol.md [hunk 24]
```diff
@@ -328,4 +498,20 @@ ex.forward()
 ex.outputs[0].asnumpy()
 ```
 
+## How Efficient Is the Symbolic API?
+
+In short, it is designed to be very efficient in both memory and runtime.
+The major reason for introducing the Symbolic API is to bring the efficient C++
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: This section is confusing a number of concepts. this sentence inparticular doesn't make much sense.

**Comment by zackchase** [comment]
Good advice. I'll revise and push an update.

File: docs/tutorials/basic/symbol.md [hunk 25]
```diff
@@ -1,23 +1,73 @@
...
-is the `Symbol` provided by `mxnet.symbol`, or `mxnet.sym` for short. A symbol
-represents a multi-output symbolic expression. They are composited by operators,
-such as simple matrix operations (e.g. "+"), or a neural network layer
-(e.g. convolution layer). An operator can take several input variables, produce
-more than one output variables, and have internal state variables. A variable
-can be either free, which we can bind with value later, or an output of another
-symbol.
-
-## Symbol Composition
+In a [previous tutorial](ndarray.md), we introduced `NDArray`,
```
Comment (mli) [inline docs/tutorials/basic/symbol.md]: that link may not work on either jupyter notebook and html. need to update `mxdoc.py` to make it work

File: docs/tutorials/basic/symbol.md (hunk 24 above)
Comment (mli) [inline docs/tutorials/basic/symbol.md]: let's remove this section. a tutorial only needs to cover the usage

**Comment by Roshrini** [comment]
Addressed all the comments.

File: docs/tutorials/basic/symbol.md [hunk 26]
```diff
@@ -1,23 +1,72 @@
...
+With symbolic programing, rather than executing operations step by step,
+we first define a *computation graph*.
+This graph contains placeholders for inputs and designated outputs.
+We can then compile the graph, yielding a function
+that can be bound to `NDArray`s and run.
+MXNet's Symbol API is similar to the network configurations
+used by [Caffe](http://caffe.berkeleyvision.org/)
+and the symbolic programming in [Theano](http://deeplearning.net/software/theano/).
+
+There are a few reasons for running symbolic operations.
```
Comment (piiswrong) [inline docs/tutorials/basic/symbol.md]: I think this entire paragraph can be removed since taking derivative is not unique to symbolic API.

mx.contrib.autograd does the same thing imperatively with NDArray

**Comment by piiswrong** [comment]
Looks good to me after removing the derivative paragraph

**Comment by Roshrini** [comment]
Removed the derivative paragraph.

File: python/mxnet/io.py [hunk 27]
```diff
@@ -487,32 +487,26 @@ class NDArrayIter(DataIter):
     >>> labels = np.ones([10, 1])
     >>> dataiter = mx.io.NDArrayIter(data, labels, 3, True, last_batch_handle='discard')
     >>> for batch in dataiter:
-    ...    print batch.data[0].asnumpy()
-    ...    batch.data[0].shape
+    >>>         print batch.data[0].asnumpy()
```
Comment (mli) [inline python/mxnet/io.py]: why change from ... to >>>? I think the former is the correct version for docstring codes

File: python/mxnet/io.py [hunk 28]
```diff
@@ -487,32 +487,26 @@ class NDArrayIter(DataIter):
...
      [[ 24.  25.]
       [ 26.  27.]]]
     (3L, 2L, 2L)
     [[[  8.   9.]
       [ 10.  11.]]
-
      [[ 20.  21.]
       [ 22.  23.]]
-
      [[ 28.  29.]
```
Comment (mli) [inline python/mxnet/io.py]: not sure if you want to remove the empty line. it may fail the future docstring test

Comment (Roshrini) [inline python/mxnet/io.py]: It doesn't render well. See the current version: http://mxnet.io/api/python/io.html#mxnet.io.NDArrayIter
After removing lines, 
screen shot 2017-05-19 at 1 47 26 pm

File: python/mxnet/io.py (hunk 27 above)
Comment (Roshrini) [inline python/mxnet/io.py]: I changed it because copy code was not working with ...
Talked to Yao, he changed "copy code" script to work with ...
So, will revert this change.

**Comment by Roshrini** [comment]
@piiswrong I don't think this build failure is caused by changes done by this PR. Can you merge this one?
